src/main/java/org/kiwiproject/dropwizard/util/startup/StartupLocker.java

```diff
@@ -0,0 +1,136 @@
+        final CuratorFramework client;
+        final InterProcessLock lock;
+        final String lockPath;
+    }
+
+    /**
+     * Attempts to acquire a lock from ZooKeeper during startup.
+     *
+     * @param lockPath      the path in ZooKeeper to store the lock
+     * @param lockTimeout   the amount of time to wait for the lock to be acquired
+     * @param curatorConfig the Curator configuration
+     * @param environment   the Dropwizard environment
+     * @return information about the acquired lock or null if lock could not be acquired
+     */
+    public StartupLockInfo acquireStartupLock(String lockPath, Duration lockTimeout, PortAssignment assignment, CuratorConfig curatorConfig, Environment environment) {
+        if (assignment == PortAssignment.STATIC) {
+            return null;
+        }
+
+        if (zkAvailabilityChecker.anyZooKeepersAvailable(curatorConfig)) {
+            var curatorFramework = curatorFrameworkHelper.startCuratorFramework(curatorConfig);
+            var lock = curatorLockHelper.createInterProcessMutex(curatorFramework, lockPath);
+
+            if (tryAcquireStartupLock(lock, lockPath, lockTimeout)) {
+                environment.lifecycle().addLifeCycleListener(new StartupWithLockJettyLifeCycleListener(curatorFramework, lock, lockPath, executioner));
+                return new StartupLockInfo(curatorFramework, lock, lockPath);
+            } else {
+                LOG.warn("Lock on path [{}] not obtained. Closing Curator.", lockPath);
+                curatorFrameworkHelper.closeQuietly(curatorFramework);
+            }
+        } else {
+            LOG.warn("No ZooKeepers are available from connect string [{}]", curatorConfig.getZkConnectString());
+        }
+
+        LOG.warn("Startup using dynamic ports will continue without ZooKeeper lock (which may result in port conflicts)");
+        return null;
+    }
+
+    private boolean tryAcquireStartupLock(InterProcessMutex lock, String lockPath, Duration lockTimeout) {
+        try {
+            LOG.debug("Start lock acquisition for path [{}]. Timeout set to [{}]", lockPath, lockTimeout);
+            curatorLockHelper.acquire(lock, lockTimeout.getQuantity(), lockTimeout.getUnit());
+            LOG.debug("Acquired lock on path [{}]", lockPath);
+            return true;
+        } catch (LockAcquisitionException e) {
+            LOG.warn("Failed to obtain startup lock. Allow startup to continue and maybe we will be lucky with port assignment!", e);
+            return false;
+        }
+    }
+
+    /**
+     * Adds a {@link StartupJettyLifeCycleListener} in case the lock was never acquired.
+     *
+     * @param lockInfo      the lock info indicating if the lock was acquired
+     * @param environment   the Dropwizard environment used to add the listener
+     */
+    public void addFallbackJettyStartupLifeCycleListener(StartupLockInfo lockInfo, Environment environment) {
+        if (isNull(lockInfo)) {
```

Comment:
If we change the `acquireStartupLock` to return an `Optional` then code that calls this (and all other) methods that take a `StartupLockInfo` would need to do a `.orElse(null)` (unless we want to have `Optional` parameters which In don't think we do)